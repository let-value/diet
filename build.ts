import { build, write, Glob, readableStreamToText, BuildOutput } from "bun";
import path from "path";
import { recipePlugin } from "./recipePlugin";

export const DIST_PATH = "dist";

async function buildTypings() {
	const { stdout } = Bun.spawn(["tsc"]);
	const text = await readableStreamToText(stdout);
	console.log(text);
}

function getEntryPoints() {
	const glob = new Glob("./src/recipes/*.tsx");
	const recipes = glob.scanSync({ cwd: "./" });

	return ["./src/index.ts", ...recipes];
}

async function buildDist(entrypoints: string[]) {
	const buildOutput = await build({
		entrypoints,
		plugins: [recipePlugin],
		target: "node",
	});

	return buildOutput;
}

async function writeDist(buildOutput: BuildOutput) {
	const result: BlobPart[] = [];

	for (const output of buildOutput.outputs) {
		result.push(await output.arrayBuffer());
	}

	await write(path.join(DIST_PATH, "index.js"), result);
}

export async function buildCookbook() {
	try {
		await buildTypings();
		const entrypoints = getEntryPoints();
		const outputs = await buildDist(entrypoints);

		if (outputs.success) {
			console.log("Build successful!");
			await writeDist(outputs);
		} else {
			console.error("Build failed!");
			console.log(outputs);
		}

		return outputs.success;
	} catch (e) {
		console.error(e);
	}
}

if (import.meta.main) {
	const success = await buildCookbook();
	process.exit(success ? 0 : 1);
}
