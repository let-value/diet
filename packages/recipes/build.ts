import { build, write, Glob } from "bun";
import path from "path";
import { recipePlugin } from "./recipePlugin";

const glob = new Glob("./src/recepies/*");
const recepies = glob.scanSync({ cwd: "./" });

const entrypoints = ["./src/index.ts", ...recepies, "./src/cookbook.ts"];

const results = await build({
	entrypoints,
	plugins: [recipePlugin],
	target: "node",
});

if (!results.success) {
	console.log(results);
	process.exit(1);
}

const result: BlobPart[] = [];

for (const output of results.outputs) {
	result.push(await output.arrayBuffer());
}

await write(path.join("dist", "recipes.js"), result);
