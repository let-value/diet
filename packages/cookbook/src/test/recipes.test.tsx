import path from "node:path";
import { writeFile, unlink } from "node:fs/promises";
import { describe, test, expect } from "bun:test";
import os from "os";
import * as everything from "@/.";

import { getEntryPoints, buildDist } from "../../build";

const entrypoints = getEntryPoints();
const dist = await buildDist(entrypoints);
const blobs: string[] = [];

for (const output of dist.outputs) {
	blobs.push(await output.text());
}

test("blobs", () => {
	expect(blobs).toHaveLength(4);
});

const temp_file = path.join(os.tmpdir(), "temp.js");

try {
	const code = blobs.join("\n");
	await writeFile(temp_file, code);

	const { recipes, buildCookbook, normalizeRecipe }: typeof everything =
		await import(temp_file);

	describe("recipes", () => {
		test("names", () => {
			const names = Object.keys(recipes);
			expect(names).toHaveLength(3);
			expect(names).toMatchSnapshot();
		});

		test("normalization", () => {
			const names = Object.values(recipes);
			const normalized = names.map(normalizeRecipe);
			expect(normalized).toHaveLength(3);
		});
	});

	describe("cookbook", () => {
		const cookbook = buildCookbook();
	});
} finally {
	await unlink(temp_file);
}
