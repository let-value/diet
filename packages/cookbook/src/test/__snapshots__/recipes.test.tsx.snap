// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`recipes names 1`] = `
[
  "Egg and Vegetable Scramble",
  "Ranch Chicken Meal Prep",
  "Chicken Stir-Fry",
]
`;
